Multiple pub subs

chris-smith opened the pull request:
- Allows multiple independent publishers/subscribers on the same topic so shutting down one will not shut down the others.
- Adds support for the `getBusInfo` XMLRPC call (this is what `rosnode info` uses)
- Adds partial support for a maximum number of XMLRPC calls before failure. Expanding this will let users
  attempt to connect to a ROS master and handle the failure cleanly on their end if there isn't one there.
- No longer supports multiple calls to `advertiseService` (mirroring `roscpp`)

`PublisherImpl` and `SubscriberImpl` are essentially moves of the previous Publisher, Subscriber code.

chris-smith committed:
Add basic support for multiple pub, subs

-Previously, multiple calls to advertise/subscribe
 would return the same publisher/subscriber so
 shutting down 1 would shutdown all of them.
-Also starts to add support for max number of
 attempts during xmlrpc calls.

chris-smith committed:
some clean up. Implement getBusInfo.

chris-smith committed:
Shutdown logging cleanup on shutdown. Add tests for multiple pub/subs. Don't advertise the same service twice per node.

IanTheEngineer commented on src/lib/Subscriber.js:
```diff
@@ -17,294 +17,89 @@
   /**
-   * Send a topic request to each of the publishers we haven't connected to yet
-   * @param pubs {Array} array of uris of nodes that are publishing this topic
+   * Get the numbber of publishers currently connected to this subscriber
```
Get the *number*

IanTheEngineer commented on src/lib/impl/PublisherImpl.js:
```diff
@@ -0,0 +1,330 @@
+/*
+ *    Copyright 2016 Rethink Robotics
```
2017 copyrights for new files

IanTheEngineer reviewed (approved):
Looks great! Thank you for splitting Pubs & Subs out into impl's like the `roscpp`  and `rospy` implementations. In addition to allowing multiple clients to open/close on the same topic, I think this will make things more maintainable for the future.

chris-smith committed:
Fix stale listener bug

-Fixes an issue with stale listeners
 on pub, sub impls.
-Adds tests to check for stale listeners.

chris-smith committed:
Fix copyright, typo

IanTheEngineer reviewed (approved):
LGTM :ship:
